torchmetrics/image/map.py

```diff
@@ -0,0 +1,201 @@
+
+import torch
+from torch import Tensor
+
+from torchmetrics.metric import Metric
+from torchmetrics.utilities.imports import _PYCOCOTOOLS_AVAILABLE
+
+if _PYCOCOTOOLS_AVAILABLE:
+    from pycocotools.coco import COCO
+    from pycocotools.cocoeval import COCOeval
+else:
+    COCO, COCOeval = None, None
+
+COCO_STATS_MAP_VALUE_INDEX = 0
+COCO_STATS_MAR_VALUE_INDEX = 8
+
+
+@dataclass
+class MAPMetricResults:
+    """Dataclass to wrap the final mAP results."""
+
+    map_value: Tensor
+    mar_value: Tensor
+    map_per_class_value: List[Tensor]
+    mar_per_class_value: List[Tensor]
+
+
+class _hide_prints:
+    """Internal helper context to suppress the default output of the pycocotools package."""
+
+    def __enter__(self):
+        self._original_stdout = sys.stdout
+        sys.stdout = open(os.devnull, "w")
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        sys.stdout.close()
+        sys.stdout = self._original_stdout
+
+
+class MAP(Metric):
+    """Computes the Mean-Average-Precision (mAP) and Mean-Average-Recall (mAR) for object detection predictions.
+    Optionally, the mAP and mAR values can be calculated per class.
+
+    https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Mean_average_precision
+
+    Boxes and targets have to be in COCO format with the box score at the end
+    (x-top left, y-top left, width, height, score)
+
+    Note:
+        This metric is a wrapper for the pycocotools, which is a standard implementation for the mAP metric
+        for object detection.
+        https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools
+
+    Note:
+        As the pycocotools library cannot deal with tensors directly, all results have to be transfered
+        to the CPU, this might have an performance impact on your training
+
+    Args:
+        num_classes:
+            Number of classes, required for mAP values per class. default: 0 (deactivate)
+    """
+
+    def __init__(
+        self,
+        num_classes: int = 0,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+    ) -> None:
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+        )
+
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+        self.add_state("average_precision", default=torch.tensor(data=[], dtype=torch.float), dist_reduce_fx="mean")
+        self.add_state("average_recall", default=torch.tensor(data=[], dtype=torch.float), dist_reduce_fx="mean")
+        self.num_classes = num_classes
+
+        for class_id in range(num_classes):
+            self.add_state(f"ap_{class_id}", default=torch.tensor(data=[], dtype=torch.float), dist_reduce_fx="mean")
+            self.add_state(f"ar_{class_id}", default=torch.tensor(data=[], dtype=torch.float), dist_reduce_fx="mean")
```

Comment:
is it correct that the `dist_reduce_fx` is `mean`? 
In the update method the values gets concatenated from different batches so I would somewhat assume this also need to happen when we have multiple devices.

Reply:
The average precision is always calculated per sample. So for batches it's the mean over all samples. If we have multiple devices it's therefore the mean of the mean, which should be correct?